File: python/cote/LTCD/이진트리_반전.py
from __future__ import annotations
from typing import Generic, TypeVar
from collections import deque
import random

TEST_CASES = 10
MAX_DEPTH = 20
MAX_VALUE = 100

T = TypeVar("T")

class Node(Generic[T]) : 
	
	def __init__(self, value: T, left: Node[T] = None, right: Node[T] = None) :
		self.value = value
		self.left = left 
		self.right = right

	def __iter__(self) :
		queue = deque([self])

		while queue :
			cur = queue.popleft()

			if cur : 
				yield cur.value
				queue.append(cur.left)
				queue.append(cur.right)

	
	def __str__(self) :
		return str([*self])

def initialize(depth: int = 0) -> Node[T] :

	if depth >= MAX_DEPTH : 
		return None

	if depth > 0 and random.random() / depth < 1 / MAX_DEPTH :
		return None

	left = initialize(depth + 1)
	right = initialize(depth + 1)

	return Node(random.randrange(MAX_VALUE), left, right)

def solve(node: Node[T]) -> Node[T] :
	queue = deque([node])

	while queue :
		cur = queue.popleft()

		if cur: 
			cur.left, cur.right = cur.right, cur.left 

			queue.append(cur.left)
			queue.append(cur.right)

	return node

def solve2(node: Node[T]) -> Node[T] :
	stack = deque([node])

	while stack :
		cur = stack.pop()

		if cur :
			stack.append(cur.left)
			stack.append(cur.right)

			cur.left, cur.right = cur.right, cur.left

	return node

if __name__ == "__main__" :
	random.seed(43)

	for _ in range(TEST_CASES) :
		node = initialize()
		print(solve(node))

File: javascript/cote/LTCD/이진트리_반전.mjs
const TEST_CASES = 10
const MAX_VALUE = 100
const MAX_DEPTH = 20

class Node {
	constructor(value, l = null, right = null) {
		this.value = value
		this.left = l
		this.right = right
	}

	*[Symbol.iterator]() {
		const queue = [this]

		while (queue.length) {
			const node = queue.shift()
			
			if (node) {
				yield node.value
				queue.push(node.left)
				queue.push(node.right)
			} else {
				yield null
			}
		}
	}
}

function initialize(depth = 0) {

	if (depth >= MAX_DEPTH) {
		return null
	}

	if (depth > 0 && Math.random() / depth < 1 / MAX_DEPTH ) {
		return null
	}

	const value = Math.floor(Math.random() * MAX_VALUE)
	const [left, right] = [initialize(depth + 1), initialize(depth + 1)];

	return new Node(value, left, right)
}

function solve(node) {

	function dfs(cur) {
		if (!cur) {
			return null
		}

		const [left, right] = [dfs(cur.right), dfs(cur.left)]
		cur.left = left 
		cur.right = right
		
		return cur
	}

	return dfs(node)
}

function solve2(node) {
	const queue = [node]

	while (queue.length) {
		const cur = queue.shift()

		if (cur) {
			const [left, right] = [cur.right, cur.left]
			cur.left = left
			cur.right = right

			queue.push(cur.left)
			queue.push(cur.right)
		}
	}
	return node
}

for (let it = 0; it < TEST_CASES; it++) {
	const node = initialize()
	console.log(...node)
	console.log(...solve2(node))
}